Sources/MongoSwift/MongoConnectionString.swift

```diff
@@ -119,6 +119,82 @@ public struct MongoConnectionString: Codable, LosslessStringConvertible {
         }
     }
 
+    private static func percentDecode(from: String?) throws -> String? {
+        guard let from = from else {
+            return nil
+        }
+        guard let decoded = from.removingPercentEncoding else {
+            throw MongoError.InvalidArgumentError(
+                message: "Connection string contains an unescaped percent sign"
+            )
+        }
+        return decoded
+    }
+
+    private mutating func applyAuthOptions(authOptions: [String: String]) throws {
+        for (key, value) in authOptions {
+            switch key {
+            case "authSource":
+                self.credential?.source = value
+            case "authMechanism":
+                if value == "GSSAPI", self.credential?.source == self.database {
+                    self.credential?.source = nil
+                }
+                self.credential?.mechanism = MongoCredential.Mechanism(value)
+            case "authMechanismProperties":
+                var authMechanismProperties = BSONDocument()
+                for property in value.components(separatedBy: ",") {
+                    let propertyKeyValue = property.components(separatedBy: ":")
+                    guard propertyKeyValue.count == 2 else {
+                        throw MongoError.InvalidArgumentError(
+                            message: "Invalid key value pair for authMechanismProperties \(propertyKeyValue)"
+                        )
+                    }
+                    let propertyKey = propertyKeyValue[0]
+                    let propertyValue = propertyKeyValue[1]
+                    switch propertyKey {
+                    case "SERVICE_NAME", "SERVICE_REALM":
+                        if let cred = self.credential?.mechanism {
```

Comment:
what if we process the auth mechanism properties before we process the mechanism option? could we run into an issue?